lnbits/lnbits
# Extension install 02

Follow up of this PR: https://github.com/lnbits/lnbits/pull/1123
Doc: [extension-install.md](https://github.com/lnbits/lnbits/blob/extension_install_02/docs/guide/extension-install.md)
Merge too complex, moved here

### Summary
This PR allows extensions to be installed/uninstalled at runtime by an admin user.
 - there is no need for a server restart after an extension has been installed
   - the db is created/updated dynamically (`migrations.py`)
   - the endpoints and static resources are mapped dynamically
 - when an extension is uninstalled all endpoints to that extension will be blocked (including static resources like `html` files)
 - extensions can be installed/uninstalled only by admin users (listed in `LNBITS_ADMIN_USERS` env var)
 - after installation, an extension is not `active`. It must be explicitly activated by the admin user.
 - regular users will only see installed extensions that have been `activated` by an admin user
 - when an extension is uninstalled the `db` for it is NOT removed
 - [x] install from GitHub releases (see extension-install.md](https://github.com/lnbits/lnbits/blob/extension_install_02/docs/guide/extension-install.md))
 - [x] show `star count` from GitHub stars (for per-repo extensions)
 - [x] allow non-admins to view install page (but not install)
 - [ ] show warning if extension has critical bugs
 - [x] allow older versions to be installed
   - the LNbits instance can have an old version that only supports older extensions
 - [ ] add dependencies in `config.json` (???)

### Limitations
 - **GitHub API**: For unauthenticated requests, the rate limit allows for up to 60 requests per hour.
**Solution**:
  - add the `LNBITS_EXT_GITHUB_TOKEN` env variable. Specifying this variable increases the rate-limit
    - this value is not visible for the super-admin therefore SaaS  can re-use the same token for all instances
    - **Note**: tokens expire after 1 year

### Test
 - use both `poetry` and `uvicorn`
 - only admin users can install/uninstall extensions

**Install&Activate**
 - install an extension (with an admin user)
 - check that regular users cannot yet access the extension
 - `Activate` the extensions
 -  users should be able to `Enable` and use the extension
 - `Deactivate` the extension
 - users should not see the extension in the list or access any UI or endpoints (`HTTP 401`)

**Dependencies**
 - an extension can depend on other extensions. This must be specified in the `dependencies: []` fields of the repository manifest file
 - if extension `A` depends on extension `B` then 
    - `A` cannot be installed until `B` is installed
    - `B` cannot be uninstalled before `A` is uninstalled
 - [ ] specify dependencies for GitHub releases

**Container Install**
- extension installation adds new files (at `lnbits/extensions`) to the filesystem. If the app runs in a docker `container` then the container filesystem state will be recreated if the container is destroyed and recreated.
- the `zips` for installed extensions are stored on the `LNBITS_DATA_FOLDER` (which must be mapped to a volume)
   ```
    volumes:
       - ./data/extensions:/app/lnbits/data/extensions
     ```
- on restart the app checks if:
    - the extension has been installed (in the DB)
    - the zip for that extension is present (if not it re-downloads it)
    - the `zip` is installed. If not it will re-install it
    -
- to test:
  - start app in container
  - install extensions
  - destroy container
  - start app again using the same image (with and without volume mapping)
  - expected result: the previously installed extensions are present

### Extension Coding Conventions
There are some rules that extensions must follow in order to be installable

**Imports**
 - no relative import of other extensions
    - **NOK**: `from ..watchonly.crud import get_config, get_fresh_address`
    - **OK**:    `from lnbits.extensions.watchonly.crud import get_config, get_fresh_address`
    - in the future extensions should communicate EXCLUSIVELY via REST API calls (no direct imports)
 - no absolute import of own modules:
    - **NOK**: `from lnbits.extensions.satspay.crud import check_address_balance, get_charge`
    - **OK**:    `from .crud import check_address_balance, get_charge, update_charge`
